### PR TITLE
DS2-245: Support uuid in firestore-service

### PR DESCRIPTION
**Pull request by vricciardulli**
Minor refactoring as well. Added many more tests. **All passing**.

Still, I'd prefer to keep this on a separate branch for now -- that's why it is "draft", but can and should be reviewed.

We can add this version to the dev service only, where I made corresponding changes https://github.com/dealroom/data-firestore-service/pull/72.

### CODE REVIEW

File: dealroom_firestore_connector/status_codes.py [hunk 1]
```diff
@@ -1,4 +1,8 @@
-SUCCESS = 0
-ERROR = -1
-CREATED = 1
-UPDATED = 2
+from enum import Enum
+
+
+class Code(int, Enum):
```
Comment (GuillermoDR) [inline dealroom_firestore_connector/status_codes.py:4]: Minor comment but what about `StatusCode` ? just for a more descriptive name. Love this change btw.

File: dealroom_firestore_connector/__init__.py [hunk 2]
```diff
@@ -1,166 +1,210 @@
...
+    NOT_IN_DEALROOM_ENTITY_ID,
+)
+
 
 # Time to sleep in seconds when a exception occurrs until retrying
 EXCEPTION_SLEEP_TIME = 5
 
 
-def new_connection(project: str, credentials_path: str = None):
+@exc_handler
```
Comment (GuillermoDR) [inline dealroom_firestore_connector/__init__.py:30]: ❤️

File: dealroom_firestore_connector/helpers.py [hunk 3]
```diff
@@ -1,5 +1,24 @@
 import logging
 import traceback
+from uuid import UUID
+from typing import Union
+
+
+def is_valid_uuid(value: Union[str, int, None]) -> bool:
+    try:
+        UUID(hex=value, version=4)
```
Comment (GuillermoDR) [inline dealroom_firestore_connector/helpers.py:9]: Nice library 👍

File: dealroom_firestore_connector/identifier.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+from typing import Union, Optional
+from .helpers import is_valid_uuid, is_valid_id
+from .exceptions import InvalidIdentifier
+
+
+# We mark the deleted entities from DR database with -2 entity so we can easily identify them.
+DELETED_DEALROOM_ENTITY_ID = -2
+
+# We mark the entities that don't exist in DR with -1.
+NOT_IN_DEALROOM_ENTITY_ID = -1
```
Comment (GuillermoDR) [inline dealroom_firestore_connector/identifier.py:11]: Maybe we can also use a class like `Code` if that makes sense here (possible Dealroom Ids) and add the `is_valid_id` there

File: dealroom_firestore_connector/__init__.py [hunk 5]
```diff
@@ -622,25 +750,41 @@ def _validate_new_people_doc_payload(payload: dict):
...
+        if "dealroom_uuid" not in payload:
+            raise KeyError(
+                "at least one of 'dealroom_id', 'dealroom_uuid' must be present in payload"
+            )
+        else:
+            _validate_dealroom_uuid(payload["dealroom_uuid"])
+    else:
+        _validate_dealroom_id(payload["dealroom_id"])
+        if "dealroom_uuid" in payload:
+            _validate_dealroom_uuid(payload["dealroom_uuid"])
```
Comment (GuillermoDR) [inline dealroom_firestore_connector/__init__.py:762]:
```
if "dealroom_id" not in payload and "dealroom_uuid" not in payload:
    raise KeyError(
                "at least one of 'dealroom_id', 'dealroom_uuid' must be present in payload"
            )

if "dealroom_id" in payload:
    _validate_dealroom_id(payload["dealroom_id"])
if "dealroom_uuid" in payload:
    _validate_dealroom_uuid(payload["dealroom_uuid"])
```

File: tests/test_all.py [hunk 6]
```diff
@@ -1,18 +1,30 @@
-import random
+"""
```
Comment (GuillermoDR) [inline tests/test_all.py:1]: Three tests fail for me locally:

```
FAILED tests/test_all.py::test_set_history_doc_refs_empty_final_url_valid_uuid
FAILED tests/test_all.py::test_set_history_doc_refs_existing_by_url_with_wrong_dealroom_uuid
FAILED tests/test_all.py::test_set_history_doc_refs_for_deleted_company_w_uuid_2
```

File: dealroom_firestore_connector/identifier.py (hunk 4 above)
Comment (vittorio-dealroom) [reply dealroom_firestore_connector/identifier.py]: Changed in https://github.com/dealroom/data-firestore-connector/pull/31/commits/e400b92cbe14e2cbe1509fef5839cabd7728530a

File: tests/test_all.py (hunk 6 above)
Comment (vittorio-dealroom) [reply tests/test_all.py:1]: The only failing test for me was `test_set_history_doc_refs_existing_by_url_with_wrong_dealroom_uuid`.

I fixed it in https://github.com/dealroom/data-firestore-connector/pull/31/commits/570f346423c82813d411c681f88ad9b15ffee66d (please read the description in the commit to understand why that change was made)

File: dealroom_firestore_connector/__init__.py (hunk 5 above)
Comment (vittorio-dealroom) [reply dealroom_firestore_connector/__init__.py:755]: Is it the same? 🤔  my eyes are crossing 🤦

File: dealroom_firestore_connector/status_codes.py (hunk 1 above)
Comment (vittorio-dealroom) [reply dealroom_firestore_connector/status_codes.py]: Changed in https://github.com/dealroom/data-firestore-connector/pull/31/commits/23d9a5ced496605ab65ca6e3b600e7e33a8f038c

File: dealroom_firestore_connector/__init__.py (hunk 5 above)
Comment (vittorio-dealroom) [reply dealroom_firestore_connector/__init__.py:753]: Changed in https://github.com/dealroom/data-firestore-connector/pull/31/commits/70bebf4fc2e28719d6efdc98bbc2cdbde62d501b

**Review by GuillermoDR — APPROVED** [review]

**Pull request merged**